Build failed after clock changed to daylight saving time

### problem

got this exception
```
[ERROR] Tests run: 47, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.506 s <<< FAILURE! - in org.apache.cloudstack.resource.ResourceCleanupServiceImplTest
[ERROR] testCalculatePastDateFromConfig(org.apache.cloudstack.resource.ResourceCleanupServiceImplTest)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.AssertionError: expected:<10> but was:<9>
	at org.apache.cloudstack.resource.ResourceCleanupServiceImplTest.testCalculatePastDateFromConfig(ResourceCleanupServiceImplTest.java:634)
```

### versions

4.20/4.21

### The steps to reproduce the bug

_No response_

### What to do about it?

_No response_

> [@wei](https://github.com/wei), was this after or during clock change? can you reproduce?

yes, actually it happened in my laptop in the past week. I had to skip the tests during building/packaging
today the issue is gone, because 10 days has past.

If I change the days to 20, then the issue is back

https://github.com/apache/cloudstack/blob/4c31f9d533e44d5f0b7f1809886c79bcb13a1ae9/server/src/test/java/org/apache/cloudstack/resource/ResourceCleanupServiceImplTest.java#L626-L635

```
[INFO] Running org.apache.cloudstack.resource.ResourceCleanupServiceImplTest
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[ERROR] Tests run: 47, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 7.255 s <<< FAILURE! - in org.apache.cloudstack.resource.ResourceCleanupServiceImplTest
[ERROR] testCalculatePastDateFromConfig(org.apache.cloudstack.resource.ResourceCleanupServiceImplTest)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.AssertionError: expected:<20> but was:<19>
	at org.apache.cloudstack.resource.ResourceCleanupServiceImplTest.testCalculatePastDateFromConfig(ResourceCleanupServiceImplTest.java:634)

```

--- a/server/src/test/java/org/apache/cloudstack/resource/ResourceCleanupServiceImplTest.java
+++ b/server/src/test/java/org/apache/cloudstack/resource/ResourceCleanupServiceImplTest.java
@@ -17,6 +17,11 @@
 package org.apache.cloudstack.resource;
 
 import java.lang.reflect.Field;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.zone.ZoneRules;
 import java.util.ArrayList;
 import java.util.Calendar;
 import java.util.Date;
@@ -629,9 +634,23 @@ public class ResourceCleanupServiceImplTest {
         Date result = resourceCleanupService.calculatePastDateFromConfig(
                 ResourceCleanupService.ExpungedResourcesPurgeKeepPastDays.key(),
                 days);
-        Date today = new Date();
-        long diff = today.getTime() - result.getTime();
-        Assert.assertEquals(days, TimeUnit.DAYS.convert(diff, TimeUnit.MILLISECONDS));
+        Instant resultInstant = result.toInstant();
+        ZoneId systemZone = ZoneId.systemDefault();
+        ZoneRules rules = systemZone.getRules();
+        ZonedDateTime resultDateTime = resultInstant.atZone(systemZone);
+        boolean isDSTInResultDateTime = rules.isDaylightSavings(resultDateTime.toInstant());
+
+        ZonedDateTime todayDateTime = ZonedDateTime.now(systemZone);
+        boolean isDSTInTodayDateTime = rules.isDaylightSavings(todayDateTime.toInstant());
+
+        Duration duration = Duration.between(resultDateTime, todayDateTime);
+        long actualDays = TimeUnit.DAYS.convert(duration.toMillis(), TimeUnit.MILLISECONDS);
+
+        if (!isDSTInResultDateTime && isDSTInTodayDateTime) {
+            Assert.assertEquals(days - 1, actualDays);
+        } else {
+            Assert.assertEquals(days, actualDays);
+        }
     }
 
     @Test

